### ISSUE TITLE
The CTM convergence criterion is identically zero on a rank-1 corner, so a collapsed environment is certified as converged at any max_iter and any conv_tol

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#898]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

Found while root-causing the full-test CI, which has been red for 15+ consecutive days (`Full tests` matrix; the required gate runs only `pytest -m core`, so none of it blocks a merge — #740 / #805).

## The defect

`_ctm_sv_diff` (`src/tenax/algorithms/_ctm_tensor_convergence.py:544-561`) normalises the corner singular values **by their sum** and returns `max|Δ|`. The loop breaks on it at `:812-834`, comparing `C1` between sweeps.

On a **rank-1 corner** both normalised vectors are `[1, 0, …, 0]`. The metric is therefore structurally incapable of exceeding one ulp, whatever the environment is doing. Measured on a collapsed fixture: `0.0` or `1.110223e-16` at every sweep from 3 onward.

So the loop exits on its second or third sweep and reports success. It is not a slow convergence — the criterion **cannot see** the environment move.

## Measured

Dense 2×2 CTM, χ=12, `conv_tol=1e-9`, `recipe="2x2"`, on the fixture from `tests/test_ctm_direction_dependent_bonds.py:30-51`:

```
it=1  eps=n/a        rank= 3/12   E=-0.4080956748401487
it=2  eps=6.265e-01  rank= 1/12   E=-0.4181474400204971
it=3  eps=1.110e-16  rank= 1/12   E=-0.4245376562156186   <-- BREAK, returned
it=4  eps=0.000e+00  rank= 1/12   E=-0.4283406780432438
it=41 eps=0.000e+00  rank= 1/12   E=-0.4332902680558767   <-- true fixed point
```

The returned value is **8.8e-3 above the fixed point**. Forcing the sweep out to 20 gives `-0.4332898820157138`, decreasing geometrically with ratio ≈0.55, extrapolating to −0.43329027 — matching the it-41 value to ~1e-9.

**Budget- and χ-independent, exactly as the mechanism predicts.** `ctm_tensor_2site` returns bit-identical `-4.2453765621561856e-01` at:

- `max_iter` = 60, 120, 300, 400, 800
- `conv_tol` = 1e-9, 1e-12, **1e-14**
- `chi` = 8, 12, 24, 48

There is no knob a caller can turn to notice.

## Blast radius, stated precisely

This bites **only when the corner is already rank-1**, and it is the reason such an environment is never reported. A healthy state converges normally — control, production path (`_simple_update_checkerboard_sweep(D=3, 80 steps)` + `_to_physical_pair`), dense 2×2:

```
chi=12: rank 9 -> 11 -> 12/12 ; eps 5.326e-02, 2.503e-02, 4.589e-03, 8.811e-04,
                                    1.206e-04, 1.842e-05, 2.482e-06  (monotone)
chi=24: rank 24/24            ; eps 5.959e-02 -> 1.608e-04
```

So this is not "CTM is broken". It is that **the collapse detector and the convergence detector are the same quantity, and it fails open on exactly the case it needs to catch** — the #723 / #726 / #747 / #666 family. A caller who hits a collapsed environment gets a converged-looking answer with no warning.

## The detectors already exist and are never called

`env_is_collapsed` (`src/tenax/algorithms/_ctm_diagnostics.py:364`) and `ctm_corner_rank` (`:350`) do exactly the right check. `grep -rn "env_is_collapsed\|ctm_corner_rank" src/` finds **no caller anywhere outside that module**.

## How it surfaces in CI

Two tests, on all three platforms, since 2026-08-12:

- `tests/test_ctm_670_symmetric_2x2.py::test_dense_2x2_energy_unchanged_by_fix`
- `tests/test_ctm_direction_dependent_bonds.py::test_symmetric_2site_ctm_matches_dense_on_direction_dependent_bonds`

**The apparent platform dependence is not BLAS.** The three distinct CI numbers are sweeps 2, 3 and 4 of the single deterministic trace above:

| CI value | platform | sweep | \|diff\| |
|---|---|---|---|
| −0.4181474400204973 (sym) | all three | 2 | 2.22e-16 |
| −0.42453765621561856 | ubuntu-3.12 | 3 | 5.55e-17 |
| −0.4283406780432455 / …4394 | ubuntu-3.11, macOS | 4 | ≤1.67e-15 |

The fixed point is identical everywhere. What varies is the exit sweep. Consequently `abs(E_sym - E_dense) < 1e-6` (`tests/test_ctm_direction_dependent_bonds.py:91`) is measuring *which of two identical runs exited first* — the symmetric corner reaches rank 1 one sweep earlier because `_corner_singular_values` (`:586-605`) is a different quantity for the two types (per-sector SVDs concatenated vs one dense χ×χ SVD). Per sweep the two paths agree to ≤4.7e-14.

**Not established:** *why* the exit sweep differs across platforms. Relative perturbations of 1e-16, 1e-15, 1e-14 and 1e-12 on the input tensors across 4 seeds (17 runs) never moved the exit sweep off 3 on this box. BLAS reassociation acting inside a sweep is a plausible mechanism but is unproven; do not repeat "eps rounds below conv_tol one sweep sooner" as established, since sweep-3 eps here is already 1.1e-16 and would need to be ≥1e-9 on those platforms.

## Suggested direction

1. **Make the criterion refuse to certify a rank-deficient corner.** Calling `env_is_collapsed` / `ctm_corner_rank` from the loop and failing closed is the obvious move; a criterion that is a no-op on the degenerate case is worse than no criterion, because it is *reported* as convergence.
2. Consider whether normalising by the sum is right at all — it discards exactly the information that distinguishes a rank-1 corner from a healthy one.
3. The two tests above will need their fixtures fixed independently (filed separately), but note they cannot be repaired by re-freezing the constant: any number derived from this path is a snapshot of a sweep index.

## Related

#723, #726, #747, #666 (rank-1 environment collapse), #780 (`conv_method='elementwise'` never converges on a physical D=4 state), #805 (both affected files are outside the required gate — `pytest -m core --collect-only` on them returns "no tests collected (5 deselected)").

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

Reopening: #903 merged as `66232b0` **without its last commit**, so one of this issue's fixes shipped nowhere.

#903's squash was taken at `66232b0` (2026-08-27T00:45:40Z). `e0a6e12` landed on the branch at 02:47:19Z and is not an ancestor of `main`. Verified: `_contributors` appears **0 times** in all four affected modules on `main`.

**The defect still live on `main`**, raised by Codex against #903 at 02:50:15Z:

> In the `2x2` recipe the top phase writes destination `c`'s C1 from `double_layers[neighbors[c]["top"]]`; a `D=1` destination fed by a nontrivial source therefore receives `max_rank=1`, allowing its collapsed rank-one corner to certify prematurely.

That is this issue's own defect reintroduced by its fix — the rank bound keyed to the coordinate a corner is *stored under*, when in the `2x2` recipe a corner is built from a **neighbour's** tensor.

**#916** carries the fix onto current `main`. Closing this again should wait for that to merge.

I should also record how it slipped: when the finding arrived I checked it against my branch head, saw the reviewed squash predated my fix, and called it stale. That was right about the branch and irrelevant to what shipped — I never checked it against the merge target. The squash-vs-head gap is generic, not specific to this PR.

**Comment by yingjerkao** [comment]
Closing: the follow-up fix **is** on `main`, and the marker used to reopen this was a false negative.

## Why the reopen check misfired

The reopening comment concluded the fix was lost because `_contributors` appears 0 times in the four affected modules. It still does — but `#916` did not ship a `_contributors` helper. It shipped a different, deliberately more conservative fix, because `neighbors` is not in scope at the convergence call site:

```
# a collapsed corner.  `neighbors` is not in scope here, so this
# coordinate's contributors, and therefore conservative: a larger
```

So the symbol was the wrong probe for the behaviour. Checking for an implementation detail rather than the defect is the same shape of error as the original one this issue is about.

## What is actually on `main`

`#916` merged as **`a613de8`** (`fix(#898): bound every site that can build a corner (follow-up, lost from #903's squash)`). The branch head `ffda75a` is not an ancestor of `main`, but that is ordinary squash-merge mechanics, not loss — I checked the squash's *content*, not its state, since "the PR is merged" is precisely the signal that failed for #903.

It carries the neighbour-source reasoning at three call sites:

```
# *neighbour's* double layer (`s_src = neighbors[s_dst]["top"]`), so
# `c` alone gives a D=1 destination fed by a rich source `max_rank=1` --
```

which is the exact defect Codex raised against #903.

## Verified by running it

```
$ pytest tests/test_ctm_criterion_rank_blind_898.py -k "richer_neighbour or global_bound_per_cell"
5 passed, 49 deselected in 12.28s
```

against `origin/main` (`ad9c213`). `test_a_coordinate_fed_by_a_richer_neighbour_is_still_covered` is the named regression for the reopening defect, and `test_one_global_bound_per_cell_can_never_under_cover` pins the conservative bound that replaced the per-coordinate one.

The original defect's guard is also present: `_ctm_sv_diff` returns `inf` on an uninformative spectrum, via `jnp.where` rather than a Python branch so it is tracer-safe inside `lax.while_loop`, with an empty-spectrum arm added by #903's own review.

Both halves are on `main` and covered. Closing.

> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.
